### PR TITLE
INTAP-85: Issue entity, bundle structure.

### PR DESCRIPTION
**Pull request by Florin1**

### CODE REVIEW

File: src/Academic/Bundle/BugTrackingBundle/Controller/IssueController.php [hunk 1]
```diff
@@ -0,0 +1,120 @@
+<?php
+
+namespace Academic\Bundle\BugTrackingBundle\Controller;
+
+use Academic\Bundle\BugTrackingBundle\Form\Handler\IssueHandler;
```
Comment (AlexPuscas) [inline src/Academic/Bundle/BugTrackingBundle/Controller/IssueController.php]: Please reorder the dependencies.

``` php
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Route;
use Sensio\Bundle\FrameworkExtraBundle\Configuration\Template;

use Symfony\Bundle\FrameworkBundle\Controller\Controller;
use Symfony\Component\HttpFoundation\RedirectResponse;

use Academic\Bundle\BugTrackingBundle\Entity\Issue;
use Academic\Bundle\BugTrackingBundle\Form\Handler\IssueHandler; 
use Academic\Bundle\BugTrackingBundle\Form\Type\IssueType;
```

File: src/Academic/Bundle/BugTrackingBundle/Controller/IssueController.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+ */
+class IssueController extends Controller
+{
+    /**
+     * @Route(
+     *      name="academic_bug_tracking_index",
+     * )
+     * @Template
+     */
+    public function indexAction()
```
Comment (AlexPuscas) [inline src/Academic/Bundle/BugTrackingBundle/Controller/IssueController.php:27]: All methods with parameters should have @param docblock.
All methods that return something should have @return docblock.

File: src/Academic/Bundle/BugTrackingBundle/Controller/IssueController.php [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+    /**
+     * @Route("/view/{id}", name="academic_bug_tracking_view", requirements={"id"="\d+"})
+     * @Template()
+     * @param Issue $issue
+     * @return array
+     */
+    public function viewAction(Issue $issue)
+    {
+        return [
```
Comment (AlexPuscas) [inline src/Academic/Bundle/BugTrackingBundle/Controller/IssueController.php]: This return can be compressed to one line.

File: src/Academic/Bundle/BugTrackingBundle/DependencyInjection/Configuration.php [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+{
+    /**
+     * {@inheritdoc}
+     */
+    public function getConfigTreeBuilder()
+    {
+        $treeBuilder = new TreeBuilder();
+        $rootNode = $treeBuilder->root('academic_bug_tracking');
+
+        // Here you should define the parameters that are allowed to
```
Comment (AlexPuscas) [inline src/Academic/Bundle/BugTrackingBundle/DependencyInjection/Configuration.php]: Is this comment required?

File: src/Academic/Bundle/BugTrackingBundle/Resources/config/importexport.yml [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+            - [setDataConverter, ['@academic.bug_tracking.importexport.data_converter']]
+            - [setStrategy, ['@academic.bug_tracking.importexport.strategy.issue.add_or_replace']]
+        tags:
+            - { name: oro_importexport.processor, type: import, entity: '%academic.bug_tracking.issue.class%', alias:  academic.bug_tracking_issue }
+            - { name: oro_importexport.processor, type: import_validation, entity: '%academic.bug_tracking.issue.class%', alias:  academic.bug_tracking_issue }
+
+     academic.bug_tracking.importexport.template_fixture.issue:
+         class: '%academic.bug_tracking.importexport.template_fixture.issue.class%'
+         tags:
+             - { name: oro_importexport.template_fixture }
```
Comment (AlexPuscas) [inline src/Academic/Bundle/BugTrackingBundle/Resources/config/importexport.yml]: Please add a new line at the end of all files, even the .yml ones.

File: src/Academic/Bundle/BugTrackingBundle/Form/EventListener/IssueSubscriber.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\Form\FormFactoryInterface;
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Bridge\Doctrine\Form\Type\EntityType;
+
+use Doctrine\ORM\EntityRepository;
+
+use Oro\Bundle\UserBundle\Entity\User;
```
Comment (dandries) [inline src/Academic/Bundle/BugTrackingBundle/Form/EventListener/IssueSubscriber.php:11]: Separate different vendors.

File: src/Academic/Bundle/BugTrackingBundle/Migrations/Data/ORM/LoadIssueData.php [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+            'default' => false
+        ],
+        Issue::STATUS_REOPENED => [
+            'label' => 'Reopened',
+            'priority' => 5,
+            'default' => false
+        ],
+    ];
+
+    public function load(ObjectManager $manager)
```
Comment (dandries) [inline src/Academic/Bundle/BugTrackingBundle/Migrations/Data/ORM/LoadIssueData.php:126]: Add phpdoc.

File: src/Academic/Bundle/BugTrackingBundle/Migrations/Schema/v1_1/AcademicBugTrackingBundle.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
+<?php
+
+namespace Academic\Bundle\BugTrackingBundle\Migrations\Schema\v1_1;
+
+use Doctrine\DBAL\Schema\Schema;
+use Oro\Bundle\MigrationBundle\Migration\Migration;
```
Comment (dandries) [inline src/Academic/Bundle/BugTrackingBundle/Migrations/Schema/v1_1/AcademicBugTrackingBundle.php:7]: Separate vendors.

File: src/Academic/Bundle/BugTrackingBundle/Migrations/Schema/v1_1/AcademicBugTrackingBundle.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+use Oro\Bundle\MigrationBundle\Migration\QueryBag;
+use Oro\Bundle\EntityExtendBundle\Migration\Extension\ExtendExtension;
+use Oro\Bundle\EntityExtendBundle\Migration\Extension\ExtendExtensionAwareInterface;
+use Oro\Bundle\EntityExtendBundle\EntityConfig\ExtendScope;
+use Academic\Bundle\BugTrackingBundle\Entity\Issue;
```
Comment (dandries) [inline src/Academic/Bundle/BugTrackingBundle/Migrations/Schema/v1_1/AcademicBugTrackingBundle.php:13]: Separate vendors.

File: src/Academic/Bundle/BugTrackingBundle/Resources/translations/messages.en.yml [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+            tags:
+                label: 'tags'
+        block:
+            general: 'General Information'
+        grid:
+            actions:
+                view:
+                    label: 'View'
+                edit:
+                    label: 'Edit'
```
Comment (dandries) [inline src/Academic/Bundle/BugTrackingBundle/Resources/translations/messages.en.yml]: Add new line.

File: src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/index.html.twig [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+    importTitle: 'Import',
+    dataGridName: gridName
+    } %}
+    <div class="btn-group pull-right">
+        {{ UI.addButton({
+            'path': path('academic_bug_tracking_create'),
+            'entity_label': 'academic.bugtracking.issue.entity_label'|trans
+        }) }}
+    </div>
+{% endblock %}
```
Comment (dandries) [inline src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/index.html.twig]: Add new line.

File: src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/searchResult.html.twig [hunk 12]
```diff
@@ -0,0 +1,16 @@
...
+{% set recordUrl = indexer_item.recordUrl %}
+
+{% set entityInfo = [
+{'title': 'academic.bugtracking.issue.created_at.label'|trans, 'value': entity.createdAt ? entity.createdAt|oro_format_datetime : 'N/A'|trans},
+{'title': 'academic.bugtracking.issue.updated_at.label'|trans, 'value': entity.updatedAt ? entity.updatedAt|oro_format_datetime : 'N/A'|trans},
+{'title': 'academic.bugtracking.issue.type.label'|trans, 'value': entity.type ? entity.type : 'N/A'|trans},
+{'title': 'academic.bugtracking.issue.resolution.label'|trans, 'value': entity.resolution ? entity.resolution : 'N/A'|trans},
+{'title': 'academic.bugtracking.issue.reporter.label'|trans, 'value': entity.reporter ? entity.reporter : 'N/A'|trans},
+{'title': 'academic.bugtracking.issue.assignee.label'|trans, 'value': entity.assignee ? entity.assignee : 'N/A'|trans},
+] %}
```
Comment (dandries) [inline src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/searchResult.html.twig]: Add new line.

File: src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/update.html.twig [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    }]
+    }] %}
+
+    {% set data = {
+    'formErrors': form_errors(form),
+    'dataBlocks': dataBlocks
+    } %}
+
+    {{ parent() }}
+{% endblock content_data %}
```
Comment (dandries) [inline src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/update.html.twig]: Add new line.

File: src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/view.html.twig [hunk 14]
```diff
@@ -0,0 +1,94 @@
...
+    'subblocks': [
+    {'data' : [generalInformation]}
+    ]
+    }
+    ] %}
+
+    {% set data = {'dataBlocks': dataBlocks} %}
+
+    {{ parent() }}
+{% endblock content_data %}
```
Comment (dandries) [inline src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/view.html.twig]: Add new line.

**Comment by AlexPuscas** [comment]
Make sure you can only create a Subtask to a Story

**Comment by AlexPuscas** [comment]
Do you really need DependencyInjection\Configuration?

File: src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Controller/IssueControllerTest.php [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+        $form['academic_bug_tracking_issue[summary]'] = '2014-03-04T20:00:00+0000';
+
+        $this->client->followRedirects(true);
+        $crawler = $this->client->submit($form);
+        $result = $this->client->getResponse();
+        $this->assertHtmlResponseStatusCodeEquals($result, 200);
+        $this->assertContains("Successfully saved issue", $crawler->html());
+    }
+
+    public function testUpdateAction()
```
Comment (AlexPuscas) [inline src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Controller/IssueControllerTest.php:45]: You can pass the ID of the issue from one test to another, and so perform all CRUD actions for the same issue, by returning the ID and depending on the previous test.

File: src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Controller/IssueControllerTest.php [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        $result = reset($result['data']);
+
+        $this->client->request(
+            'GET',
+            $this->getUrl('academic_bug_tracking_view', array('id' => $result['id']))
+        );
+        $result = $this->client->getResponse();
+
+        $this->assertHtmlResponseStatusCodeEquals($result, 200);
+        $this->assertContains('General Information', $result->getContent());
```
Comment (AlexPuscas) [inline src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Controller/IssueControllerTest.php]: You should test something specific for an issue, like name or code.

File: src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Entity/Repository/IssueRepositoryTest.php [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+            ->getContainer()
+            ->get('doctrine')
+            ->getRepository(Issue::class);
+    }
+
+    public function testGetIssuesGroupedByStatus()
+    {
+        $result = reset($this->repository->getIssuesGroupedByStatus());
+
+        $this->assertArrayHasKey('status', $result);
```
Comment (AlexPuscas) [inline src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Entity/Repository/IssueRepositoryTest.php:32]: You should also check the result of the query.

File: src/Academic/Bundle/BugTrackingBundle/Tests/Unit/Form/Type/IssueTypeTest.php [hunk 18]
```diff
@@ -0,0 +1,104 @@
...
+        $this->resolver->expects($this->any())
+            ->method('setDefaults')
+            ->with($this->isType('array'));
+
+        $this->issueType = new IssueType($this->dataClass, $this->eventListener);
+    }
+
+    public function testBuildForm()
+    {
+        $this->issueType->buildForm($this->builder, []);
```
Comment (AlexPuscas) [inline src/Academic/Bundle/BugTrackingBundle/Tests/Unit/Form/Type/IssueTypeTest.php]: You should actually test the submit of the form. See http://symfony.com/doc/current/form/unit_testing.html

**Comment by AlexPuscas** [comment]
Please unit test classes from DependencyInjection folder.

File: src/Academic/Bundle/BugTrackingBundle/Form/Type/IssueType.php [hunk 19]
```diff
@@ -0,0 +1,129 @@
...
+                    'enum_code' => Issue::TYPE_ENUM_CODE,
+                    'configs' => ['allowClear' => false],
+                    'excluded_values' => [Issue::TYPE_SUBTASK],
+                    'disabled' => $disabled
+                ]
+            );
+        } else {
+            $builder->add(
+                'parent',
+                EntityType::class, array(
```
Comment (yshyshkin) [inline src/Academic/Bundle/BugTrackingBundle/Form/Type/IssueType.php]: Please, maintain consistency and use short syntax array all over the code.

**Comment by yshyshkin** [comment]
Please, fix code style issues:

```
maniac@laphere:/opt/www/academ-fo$ phpcs --encoding=utf-8 --extensions=php --standard=psr2 src/ -p
..........EE....E..............


FILE: ...fo/src/Academic/Bundle/BugTrackingBundle/Form/Type/IssueType.php
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 84 | ERROR | [x] Only one argument is allowed per line in a
    |       |     multi-line function call
 87 | ERROR | [x] Closing parenthesis of a multi-line function call
    |       |     must be on a line by itself
----------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: .../Bundle/BugTrackingBundle/Form/EventListener/IssueSubscriber.php
----------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
----------------------------------------------------------------------
 60 | ERROR | [x] Expected 0 spaces before closing bracket; newline
    |       |     found
 66 | ERROR | [x] Expected 0 spaces before closing bracket; newline
    |       |     found
----------------------------------------------------------------------
PHPCBF CAN FIX THE 2 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------


FILE: ...ackingBundle/Tests/Functional/Controller/IssueControllerTest.php
----------------------------------------------------------------------
FOUND 3 ERRORS AFFECTING 1 LINE
----------------------------------------------------------------------
 78 | ERROR | [x] Multi-line function call not indented correctly;
    |       |     expected 8 spaces but found 12
 78 | ERROR | [x] Only one argument is allowed per line in a
    |       |     multi-line function call
 78 | ERROR | [x] Closing parenthesis of a multi-line function call
    |       |     must be on a line by itself
----------------------------------------------------------------------
PHPCBF CAN FIX THE 3 MARKED SNIFF VIOLATIONS AUTOMATICALLY
----------------------------------------------------------------------
```

File: src/Academic/Bundle/BugTrackingBundle/Form/EventListener/IssueSubscriber.php [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+            );
+    }
+
+    /**
+     * @return callable
+     */
+    protected function getUsersClosure()
+    {
+        return function (EntityRepository $entityRepository) {
+            return $entityRepository->createQueryBuilder('user')
```
Comment (yshyshkin) [inline src/Academic/Bundle/BugTrackingBundle/Form/EventListener/IssueSubscriber.php]: Please, move direct query builder manipulation to entity repository and then call appropriate method here- form layer should not be aware about database structure.

File: src/Academic/Bundle/BugTrackingBundle/Placeholder/PlaceholderFilter.php [hunk 21]
```diff
@@ -0,0 +1,27 @@
...
+use Symfony\Component\HttpFoundation\Request;
+
+class PlaceholderFilter
+{
+    /**
+     * @var Request
+     */
+    protected $request;
+
+    public function __construct(Request $request)
```
Comment (yshyshkin) [inline src/Academic/Bundle/BugTrackingBundle/Placeholder/PlaceholderFilter.php:18]: Please, add missing phpdoc.

File: src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/index.html.twig [hunk 22]
```diff
@@ -0,0 +1,28 @@
...
+"%entity%": 'academic.bugtracking.issue.entity_plural_label'|trans,
+}
+}) %}
+
+{% block navButtons %}
+    {% include 'OroImportExportBundle:ImportExport:buttons.html.twig' with {
+    entity_class: entity_class,
+    exportProcessor: 'academic.bug_tracking_issue',
+    exportTemplateProcessor: 'academic.bug_tracking_issue',
+    exportTitle: 'Export',
```
Comment (yshyshkin) [inline src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/index.html.twig]: Please, translate this label.

File: src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/index.html.twig [hunk 23]
```diff
@@ -0,0 +1,28 @@
...
+    importProcessor: 'academic.bug_tracking_issue',
+    importTitle: 'Import',
```
Comment (yshyshkin) [inline src/Academic/Bundle/BugTrackingBundle/Resources/views/Issue/index.html.twig]: Please, translate this label.

File: src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Entity/Repository/IssueRepositoryTest.php (hunk 17 above)
Comment (yshyshkin) [inline src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Entity/Repository/IssueRepositoryTest.php:32]: Please, provide better assertion of method results.

**Review by yshyshkin — CHANGES_REQUESTED** [review]

File: src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Entity/Repository/IssueRepositoryTest.php [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+        $this->assertEquals($result['status'], 'open');
```
Comment (yshyshkin) [inline src/Academic/Bundle/BugTrackingBundle/Tests/Functional/Entity/Repository/IssueRepositoryTest.php:33]: Please, don't forget that first argument should contain expected value, and second argument - actual value.

File: src/Academic/Bundle/BugTrackingBundle/Placeholder/PlaceholderFilter.php [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+
+class PlaceholderFilter
+{
+    /**
+     * @var Request
+     */
+    protected $request;
+
+    /**
+     * PlaceholderFilter constructor.
```
Comment (yshyshkin) [inline src/Academic/Bundle/BugTrackingBundle/Placeholder/PlaceholderFilter.php:15]: Please, avoid useless (e.g. autogenerated) comments. Comments should provide valuable information for developer, if they don't do that - it's better to remove them. In this case would be better to have only phpdoc in method comment.

**Pull request closed without merging**